Full query response cache plugin

glasser opened the pull request:
Implements a full query cache for Apollo Server in the new package `apollo-server-plugin-response-cache`. This has similar functionality to the caching feature of the deprecated Engine proxy.

Supporting this required:

- GraphQLRequestContext new fields:
  - overallCachePolicy
  - documentText
  - metrics

- New plugin hook responseForOperation.

- new GraphQLExtension hook didResolveOperation, identical to the same hook in
  the Plugin API.  Change apollo-engine-reporting to use this hook instead of
  executionDidStart, because executionDidStart doesn't run if the cache
  short-circuits execution.

- apollo-engine-reporting: report whether the request was a cache hit. Also use
  the new requestContext.metrics object to report persisted query hit/register
  instead of specific extension options (though those extension options still
  work).

- cacheControl constructor option semantic change: include the cacheControl
  GraphQL extension in the output with `cacheControl: true` and `cacheControl:
  {stripFormattedExtensions: false}` (as before), but not for `cacheControl:
  {otherOptions: ...}`.

glasser commented:
Added a basic end-to-end test (without any of the configurable options like session ID/privacy).

martijnwalraven commented on packages/apollo-server-core/src/requestPipeline.ts:
```diff
@@ -290,26 +290,47 @@ export async function processGraphQLRequest<TContext>(
       >,
-    );
+    )).filter(x => x);
+    if (pluginGraphQLResponses.length > 1) {
```
I'm worried enforcing the rule that only one plugin can return a response will be too restrictive. I can even imagine wanting to add multiple caching plugins. An alternative would be to invoke hooks in order (instead of in parallel) until a plugin returns a response. (Maybe we could add a `Dispatcher.invokeHookSequentially`?)

martijnwalraven commented on packages/apollo-server-core/src/requestPipelineAPI.ts:
```diff
@@ -57,6 +67,8 @@ export interface GraphQLRequestContext<TContext = Record<string, any>> {
   readonly operationName?: string | null;
   readonly operation?: OperationDefinitionNode;
 
+  readonly overallCachePolicy?: Required<CacheHint> | undefined;
```
Because `CacheHint`, `CacheScope`, and `overallCachePolicy` are tied to our caching spec, it might be better to leave them in `apollo-cache-control`. We can still augment `GraphQLRequestContext` with `overallCachePolicy` there.

glasser replied:
What do you mean by augment? Is that a TypeScript feature I'm not familiar with? I think I'm not following this suggestion.

martijnwalraven replied:
Yep, it's called 'module augmentation' and leads to merging of the declarations: https://www.typescriptlang.org/docs/handbook/declaration-merging.html

So you should be able to add something like this to `apollo-cache-control`:
```
declare module 'apollo-server-core/dist/requestPipelineAPI' {
  interface GraphQLRequestContext<TContext> {
    readonly overallCachePolicy?: Required<CacheHint> | undefined;
  }
}
```

As a side note, we've been thinking about extracting `requestPipelineAPI` into its own `apollo-server-api` package to avoid circular dependencies.

martijnwalraven commented on packages/apollo-server-plugin-base/src/index.ts:
```diff
@@ -32,12 +32,23 @@ export interface GraphQLRequestListener<TContext = Record<string, any>> {
+  // GraphQLResponse, that result is used instead of executing the query. It is
+  // an error for more than one plugin to return a value that resolves to a
+  // non-null GraphQLResponse.
+  execute?(
```
I agree with the idea of adding a hook like this, but the name `execute` seems confusing. This isn't really execution, but rather a way to short-circuit execution. We don't invoke `executionDidStart` when a response is returned from cache for example, because we don't measure execution time in that case. Maybe something like `responseForOperation`?

martijnwalraven replied:
There is also possible confusion/overlap with the addition of a pluggable executor, which was added in https://github.com/apollographql/apollo-server/commit/185dd45f4360d10fb50801b7fbb26df98db4d19b and is used to plug in Apollo Gateway. I think we want to keep this separate, both because the gateway isn't a regular plugin (it composes a schema from a service list and it needs to reload itself when service list changes) and because we actually want to be able to cache gateway responses.

martijnwalraven commented on packages/apollo-server-plugin-full-query-cache/src/ApolloServerPluginFullQueryCache.ts:
```diff
@@ -0,0 +1,273 @@
+
+  // If this hook is defined and returns false, the plugin will not read
+  // responses from the cache.
+  readFromCache?(
```
This sounds too much like an action. Maybe `shouldReadFromCache`?

martijnwalraven commented on packages/apollo-server-plugin-full-query-cache/src/ApolloServerPluginFullQueryCache.ts:
```diff
@@ -0,0 +1,273 @@
+
+  // If this hook is defined and returns false, the plugin will not write the
+  // response to the cache.
+  writeToCache?(
```
Maybe `shouldWriteToCache`?

martijnwalraven commented on packages/apollo-server-plugin-full-query-cache/src/ApolloServerPluginFullQueryCache.ts:
```diff
@@ -0,0 +1,273 @@
+}
+
+interface BaseCacheKey {
+  document: string;
```
A `document` property containing the document text seems confusing, because we use `document` on the request context to refer to the parsed `DocumentNode`.

martijnwalraven commented on packages/apollo-server-plugin-full-query-cache/src/ApolloServerPluginFullQueryCache.ts:
```diff
@@ -0,0 +1,273 @@
+          baseCacheKey = {
+            // XXX could also have requestPipeline add the unparsed document to requestContext;
+            // can't just use requestContext.request.query because that won't be set for APQs
+            document: print(requestContext.document),
```
I'd really like to avoid having to print the document AST every time we need to compute a cache key. So the suggestion of putting the document text on the request context makes sense to me.

martijnwalraven replied:
Alternatively, we could replace the use of `document` and `operationName` by an `operationID`, which would rely on the same normalization we use for the operation registry. So we could add an `operationID` to the request context and use that across plugins.

glasser replied:
Gosh, the op reg stuff is so in flux though. Are you imagining this ID as just the pair of normalized doc and ID, or as an already hashed thing? I'm sort of not psyched about 100% standardizing on a cross-codebase normalization until  https://github.com/graphql/graphql-js/pull/1628 is merged. Though I also think it should be kosher to change cache key calculation later if we need. (Down for adding the string to the context though.)

martijnwalraven replied:
I was imagining the operation ID as a hash based on the normalized operation node + dependent fragments (so not the entire document, in case that contains multiple operations). I agree we want to use `stripIgnoredTokens` there. Realistically though, we can't expect everyone to be on the latest `graphql` even after that PR lands. Since it's only [a small bit of code](
https://github.com/graphql/graphql-js/pull/1628/files#diff-fea7ac4cd268610ae0fa84205423b0a7), maybe we should just include it for now?

glasser replied:
That would be my plan, but I'd like to see it merged and ideally released first, and the current PR has a comment about how the implementation may change.

martijnwalraven commented on packages/apollo-server-plugin-full-query-cache/src/ApolloServerPluginFullQueryCache.ts:
```diff
@@ -0,0 +1,273 @@
+        if (value === undefined) {
+          return null;
+        }
+        return JSON.parse(value);
```
Note that this currently will not return a valid `GraphQLResponse`, because the `errors` property is expected to contain instances of `GraphQLError`, not formatted errors. I've been looking at changing this to `GraphQLFormattedError` however, because I'm running into a similar problem in the gateway where we want to propagate errors from underlying services. So this gives me another data point in favor of making that change.

martijnwalraven replied:
I'll have a look at that today, to get a better sense of how much work it would be. Ideally we'd get it done this week as well, because I also need it for the gateway work. My main concern is that it's potentially breaking for people who have already adopted the plugin API. So I want to hear what Jesse thinks about that as well.

glasser replied:
So just to be clear, the issue is that GraphQLResponse isn't actually a json roundtrippable object?

martijnwalraven replied:
Yep

martijnwalraven replied:
It's also more consistent with how the `errors` field of a response is described in the spec: 
https://facebook.github.io/graphql/draft/#sec-Errors. So the idea is that we'd add a separate `didEncounterErrors` hook for error reporting, that will receive `GraphQLError` instances. But the response would contain the errors after applying `formatError`.

martijnwalraven commented on package.json:
```diff
@@ -58,6 +58,7 @@
     "apollo-server-lambda": "file:packages/apollo-server-lambda",
     "apollo-server-micro": "file:packages/apollo-server-micro",
     "apollo-server-plugin-base": "file:packages/apollo-server-plugin-base",
+    "apollo-server-plugin-full-query-cache": "file:packages/apollo-server-plugin-full-query-cache",
```
This is a bit of a meta point, but I'm wondering if 'full query cache' is clear enough. I think we've been using 'full query cache' and 'whole response cache' interchangeably, but I think 'full query cache' is too easily confused with the APQ cache or the document cache. Maybe we should rename this to `apollo-server-plugin-response-cache` instead?

glasser commented:
Notes:
- don't write things back to cache
- make sure http cache headers are set appropriately (eg with Age)
- set trace flags

glasser commented:
- ignore mutations

glasser commented:
I renamed the package to apollo-server-plugin-response-cache. To make future rebases easier I squashed and force pushed. Other changes will happen as additional commits, though.

glasser commented:
Note that this is currently rebased on top of #2441. I think all the commits after that PR can be squashed together, but I kept them separate to make it easier to review for people who already saw the first version of this.

I think this is feature complete. It needs tests for a few more features, and needs docs. I was going to add docs to the part of our docs that already describe using `@cacheControl` for HTTP header purposes... but I can't find that? Anyone know where that ended up?

glasser commented:
I have to run for the day — the test failure is a hapi thing, I should fix that.

glasser commented:
This is no longer built on top of #2441. It doesn't use the executor interface added there, but uses a separate responseForOperation hook, at @martijnwalraven 's suggestion.

abernix commented on packages/apollo-server-plugin-base/src/index.ts:
```diff
@@ -32,6 +32,17 @@ export interface GraphQLRequestListener<TContext = Record<string, any>> {
+  responseForOperation?(
+    requestContext: WithRequired<
+      GraphQLRequestContext<TContext>,
+      'document' | 'operationName' | 'operation'
```
In the same spirit as one of my previous comments, I think `metrics` is guaranteed to be here at this point, right?  Could we declare `metrics` as required and avoid some not-null assertions? 

```suggestion
      'document' | 'operationName' | 'operation' | 'metrics'
```

glasser replied:
Honestly I'd like to declare it as required overall ... it's kinda a shame that GraphQLRequestContext has to represent both the argument to processGraphQLRequest *and* the argument to hooks it invokes. Maybe I could define some HookGraphQLRequestContext that's a WithRequired which at least has source and metrics set on it? Though maybe that's not great.

abernix replied:
We do have some progress we can make on improving this, but I'd not ask you to address it here.  There are other hooks that could benefit from something in this regard and @martijnwalraven 	and I have been musing about an `apollo-server-api` that could take the place of this entirely.

I do think some types could at least be shared between here and Apollo Server, but we want to be careful to allow plugins to have more flexibility and not be explicitly pinned to a specific Apollo Server version, thus the loose coupling and duplication.  Apollo Server could probably depend on this same type when it invokes the plugin, but that also has possibly similar undesirable couplings.  WIP, to say the least.

abernix commented on packages/apollo-server-plugin-response-cache/src/ApolloServerPluginResponseCache.ts:
```diff
@@ -0,0 +1,337 @@
+
+      return {
+        async responseForOperation(
+          requestContext: WithRequired<
```
Hmm, this should be already inferred from the `apollo-server-plugin-base` and not necessary to be explicitly defined here, no?

glasser replied:
What do you mean — when implementing a method you don't need to declare its arguments?

abernix replied:
I believe that you can just have `requestContext` here (leaving off the duplicative re-typing) and the type will be inferred as the `WithRequired<....>` type as defined in [this bit from above](https://github.com/apollographql/apollo-server/pull/2437/files#diff-0f0cc848fe2c6442424f2ea0559f35ecR40)?

abernix commented on packages/apollo-server-plugin-base/src/index.ts:
```diff
@@ -32,6 +32,17 @@ export interface GraphQLRequestListener<TContext = Record<string, any>> {
+  ): ValueOrPromise<GraphQLResponse | null>;
   executionDidStart?(
     requestContext: WithRequired<
       GraphQLRequestContext<TContext>,
```
Should we also extend the `WithRequired` here to indicate that `metrics` is available?

abernix commented on packages/apollo-server-core/src/requestPipeline.ts:
```diff
@@ -284,32 +289,53 @@ export async function processGraphQLRequest<TContext>(
+        'executionDidStart',
+        requestContext as WithRequired<
+          typeof requestContext,
+          'document' | 'operation' | 'operationName'
```
I believe `metrics` is guaranteed here?

```suggestion
          'document' | 'operation' | 'operationName' | 'metrics'
```

Also, maybe the same treatment on `didResolveOperation` above?

abernix commented on packages/apollo-server-plugin-response-cache/src/ApolloServerPluginResponseCache.ts:
```diff
@@ -0,0 +1,337 @@
+            'document' | 'operationName' | 'operation'
+          >,
+        ): Promise<GraphQLResponse | null> {
+          requestContext.metrics!.responseCacheHit = false;
```
I suspect we could get rid of the non-null `!` assertion here with the `WithRequired` changes for `metrics` I've suggested previously?

abernix commented on packages/apollo-server-plugin-response-cache/src/ApolloServerPluginResponseCache.ts:
```diff
@@ -0,0 +1,337 @@
+        },
+
+        async willSendResponse(
+          requestContext: WithRequired<GraphQLRequestContext<any>, 'response'>,
```
The type for `requestContext` should be inferred?

abernix commented on packages/apollo-server-plugin-response-cache/tsconfig.json:
```diff
@@ -0,0 +1,14 @@
+  "references": [
+    { "path": "../apollo-cache-control" },
+    { "path": "../apollo-server-plugin-base" },
+    { "path": "../apollo-server-caching" }
```
I'm somewhat surprised we don't reference the `apollo-server-env` project here, but also noticing that we don't do that anywhere else. 🤷‍♂️ 	

Note: I don't think this means you should change anything, just a casual observation when I was mentally double-clicking on the dependencies.

abernix commented on packages/graphql-extensions/src/index.ts:
```diff
@@ -108,6 +112,16 @@ export class GraphQLExtensionStack<TContext = any> {
+  }) {
+    this.extensions.forEach(extension => {
+      if (extension.didResolveOperation) {
+        extension.didResolveOperation(o);
```
I guess we don't guard on whether or not the hook is a `typeof` `===` `'function'` in other places, so this isn't the exception to the pattern.

glasser commented on packages/apollo-server-core/src/requestPipelineAPI.ts:
```diff
@@ -50,13 +60,16 @@ export interface GraphQLRequestContext<TContext = Record<string, any>> {
   readonly queryHash?: string;
 
   readonly document?: DocumentNode;
+  readonly documentText?: string;
```
That's pretty good!

abernix replied:
I think we should do it!

StefanFeederle commented:
Awesome PR! Did you run benchmarks on the reponse time before/after?

glasser commented:
Updated based on review. Finished the tests I wanted to write. Have to head out early today so I haven't done docs yet.

glasser commented:
@abernix @martijnwalraven Tests and docs are good. It would be great if you could review this today for merge tomorrow!

martijnwalraven commented on packages/apollo-server-plugin-response-cache/src/ApolloServerPluginResponseCache.ts:
```diff
@@ -0,0 +1,330 @@
+  //
+  // This hook may return a promise because, for example, you might need to
+  // validate a cookie against an external service.
+  sessionId?(
```
Should this be `sessionID`? I know the current codebase isn't very consistent, but I think we said we would follow https://github.com/airbnb/javascript#naming--Acronyms-and-Initialisms.

glasser replied:
I looked into this. apollo-server and apollo-client are almost 100% consistent in not using `ID` except stand-alone as the built-in GraphQL `ID`. This includes `clientReferenceId` and `dataIdFromObject` (an important apollo-client hook).  The only instance of `\BID` in our APIs in these two projects is `serviceID` in the plugin API, used only by the operation registry, which we should perhaps change to `serviceId` (keeping around `serviceID` for compat for a bit too).

Note that there's some code that tries to pass a `serviceID` to the EngineReportingAgent constructor too but I'm not sure why; it doesn't read it.

martijnwalraven commented on docs/source/features/caching.md:
```diff
@@ -0,0 +1,131 @@
+
+The easiest way to add cache hints is directly in your schema using the `@cacheControl` directive. Apollo Server automatically adds the definition of the `@cacheControl` directive to your schema when you create a new `ApolloServer` object with `typeDefs` and `resolvers`.
+
+You can apply `@cacheControl` to an individual field or to a type. Hints on a type apply to all fields that *return* objects of that type (not to the fields inside that type). Hints on fields override hints specified on the target type.  `@cacheControl` can specify `maxAge` (in seconds, like in an HTTP `Cache-Control` header) and `scope`, which can be `PUBLIC` (the default) or `PRIVATE`.
```
I know what you mean by "Hints on a type apply to all fields that *return* objects of that type (not to the fields inside that type)", but I think this may require some clarification for it to make sense to people.

glasser replied:
OK, I added a bunch of specific examples.

glasser commented on packages/apollo-server-core/src/requestPipeline.ts:
```diff
@@ -323,7 +353,7 @@ export async function processGraphQLRequest<TContext>(
     }
 
-    return sendResponse(response);
+    return sendResponse(response!!);
```
Heh, I think I'm getting `!` confused with another programming language.  Thanks for fixing this.

glasser commented on docs/source/features/caching.md:
```diff
@@ -0,0 +1,131 @@
+---
+title: Caching
+description: Automatically set HTTP cache headers! Save full responses in a cache!
```
I thought it parsed better than two period sentences and it made it clearer that there are two independent features than using `and` or `or` but fair enough, I changed it.

glasser commented:
@StefanFeederle I've done some benchmarking but the performance improvements will depend on your resolvers' workloads.

guoliu commented:
@glasser curious to know if you think whether this feature is a reasonable request or not 😄https://github.com/apollographql/apollo-server/issues/3228

kminehart commented:
Question about this: how would you manually invalidate the cache after a mutation that updates cached data?

I have a couple queries,

`getSingle`, `getList`, `create`, `update`.

If I `getList` and `getSingle`, then those responses are then cached. If I `update`, then `getSingle` will return stale data. Yet I don't see a way here to manually say, "invalidate the cached results for `getSingle` and `getList`."

Am I missing something in the PR here or should I make an issue?

glasser commented:
You're right that there's no invalidation support or way to get the cache key so that you can manually invalidate. #3228 is related. You are welcome to file a PR; I don't personally focus on Apollo Server these days so I'm not sure what the prioritization of such a thing is. It certainly is a major missing piece of this plugin.

kminehart commented:
Thanks! I was thinking I I was missing something, related to hinting or the Cache-Control header. I couldn't find much info elsewhere.

Thanks for the response!

guoliu commented:
> Question about this: how would you manually invalidate the cache after a mutation that updates cached data?
> 
> I have a couple queries,
> 
> `getSingle`, `getList`, `create`, `update`.
> 
> If I `getList` and `getSingle`, then those responses are then cached. If I `update`, then `getSingle` will return stale data. Yet I don't see a way here to manually say, "invalidate the cached results for `getSingle` and `getList`."
> 
> Am I missing something in the PR here or should I make an issue?

@kminehart we have the exactly same problem, and we solved it by two directives: one on queries for keeping a mapping from object global id to the list of associating cache keys, another one on mutations to purge the associating cache keys for the given global id been updated.

However, this requires exposing the cache keys in context to access it downstream. So we are currently using our own fork or plugin. More detail https://github.com/apollographql/apollo-server/issues/3228. If exposing the cache key is useful to you, we can make a PR.

kminehart commented:
Yeah we've decided to implement our own plugin instead that supports these
kinds of operations. I imagine most of our changes match the ones you did
in your fork.

On Tue, Mar 17, 2020 at 7:35 PM Guo Liu <notifications@github.com> wrote:

> Question about this: how would you manually invalidate the cache after a
> mutation that updates cached data?
>
> I have a couple queries,
>
> getSingle, getList, create, update.
>
> If I getList and getSingle, then those responses are then cached. If I
> update, then getSingle will return stale data. Yet I don't see a way here
> to manually say, "invalidate the cached results for getSingle and getList
> ."
>
> Am I missing something in the PR here or should I make an issue?
>
> @kminehart <https://github.com/kminehart> we have the exactly same
> problem, and we solved it by two directives: one on queries for keeping a
> mapping from object global id to the list of associating cache keys,
> another one on mutations to purge the associating cache keys for the given
> global id been updated.
>
> However, this requires exposing the cache keys in context to access it
> downstream. So we are currently using our own fork or plugin. More detail
> #3228 <https://github.com/apollographql/apollo-server/issues/3228>. If
> exposing the cache key is useful to you, we can make a PR.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apollographql/apollo-server/pull/2437#issuecomment-600366264>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABHHCW2QJ2DMGW6OKG3AUVDRIAJOVANCNFSM4G5QEX3Q>
> .
>
